lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -76,7 +80,20 @@ module Constants
         resource_type: 'ml_job',
         extra_common_labels: %w(job_id task_name)
       }
-      SUB_SERVICE_LIST = [DATAFLOW_CONSTANTS, ML_CONSTANTS]
+
+      # The map between a subservice name and a resource type.
+      SUBSERVICE_MAP = \
+        [APPENGINE_CONSTANTS, CONTAINER_CONSTANTS, DATAFLOW_CONSTANTS,
+         DATAPROC_CONSTANTS, ML_CONSTANTS]
+        .map { |consts| [consts[:service], consts[:resource_type]] }.to_h
+      # Default back to GCE if invalid value is detected.
+      SUBSERVICE_MAP.default = COMPUTE_CONSTANTS[:resource_type]
+
+      # The map between a resource type and expected subservice attributes.
+      SUBSERVICE_METADATA_ATTRIBUTES = \
+        [APPENGINE_CONSTANTS, CONTAINER_CONSTANTS, DATAPROC_CONSTANTS]
+        .map { |consts| [consts[:resource_type], consts[:metadata_attributes]] }
```

Comment:
If you added `.to_set` here, you could have avoided it in the constants themselves...

Reply:
Addressed in the follow up PR: https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/148/commits/f3fbf25a37cf73dc44576d57be30ea77f7a0c322.